kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,700 @@
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
+  local ipc = {
+    register_listeners = function(events)
+      -- The DNS client library will be required in globalpatches before Kong
+      -- initializes worker_events.
+      if not kong or not kong.worker_events then
+        return
+      end
+      local cwid = ngx.worker.id()
+      for _, ev in pairs(events) do
+        local handler = function(data, event, source, wid)
+          if cwid ~= wid then
+            ev.handler(data)
+          end
+        end
+        kong.worker_events.register(handler, ipc_source, ev.channel)
+      end
+    end,
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc             = ipc,
+    neg_ttl         = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    lru_size        = opts.cache_size or 10000,
+    shm_miss        = "kong_dns_cache_miss",
+    resty_lock_opts = resty_lock_opts,
+  })
+
+  if not cache then
+    return nil, "could not create mlcache: " .. err
+  end
+
+  if opts.cache_purge then
+    cache:purge(true)
+  end
+
+  -- parse order
+  local search_types = {}
+  local order = opts.order or DEFAULT_ORDER
+  local preferred_ip_type
+  for _, typstr in ipairs(order) do
+    local qtype = valid_type_names[typstr:upper()]
+    if not qtype then
+      return nil, "Invalid dns record type in order array: " .. typstr
+    end
+
+    table_insert(search_types, qtype)
+
+    if (qtype == TYPE_A or qtype == TYPE_AAAA) and not preferred_ip_type then
+      preferred_ip_type = qtype
+    end
+  end
+  preferred_ip_type = preferred_ip_type or TYPE_A
+
+  if #search_types == 0 then
+    return nil, "Invalid order array: empty record types"
+  end
+
+  -- parse hosts
+  local hosts = init_hosts(cache, opts.hosts, preferred_ip_type)
+
+  return setmetatable({
+    cache         = cache,
+    stats         = {},
+    hosts         = hosts,
+    r_opts        = r_opts,
+    resolv        = opts._resolv or resolv,
+    valid_ttl     = opts.valid_ttl,
+    error_ttl     = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl     = opts.stale_ttl or DEFAULT_STALE_TTL,
+    empty_ttl     = opts.empty_ttl or DEFAULT_EMPTY_TTL,
+    search_types  = search_types,
+  }, mt)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = errcode == NAME_ERROR_CODE and self.empty_ttl or self.error_ttl
+    -- compatible with balancer, which needs this field
+    answers.expire = now() + answers.ttl
+    return answers
+  end
+
+  local processed_answers = {}
+  local cname_answer
+
+  local ttl = self.valid_ttl or 0xffffffff
```

Comment:
What's the meaning of `0xffffffff`? Does it means `-1`?

Reply:
in the context of 32-bit unsigned integers, 0xffffffff represents the maximum value that can be stored in such a data type. When used to represent a TTL (Time to Live) value, 0xffffffff would indicate the highest possible TTL, effectively meaning that the record should not be considered stale and should remain in the cache indefinitely, or until it is explicitly removed or updated.

In Lua, if you're using this value to set a very long TTL, you might define it as a constant like this:

local MAX_TTL = 0xffffffff
This constant can then be used throughout your code to set a TTL that is as high as possible for the given system's limitations. It's important to note that while this value is the maximum for a 32-bit unsigned integer, actual TTL values in DNS and other systems are often limited to a much lower range, typically not exceeding a few years, to ensure data freshness and prevent outdated information from being cached for too long.

Reply:
comment done https://github.com/Kong/kong/pull/12305/commits/1e20d8293b93229af4f9ede801906b0fd81ecb9c

Reply:
This relies on the architecture of the system. Can't we simply specify 10*365*24*60*60 for a 10 year ttl which will fit in any architecture, and is more readable?